.github/workflows/ci.yml

```diff
@@ -157,13 +157,40 @@ jobs:
           AGENT_MODE: ${{ vars.AGENT_MODE }}
           AGENT_TYPE: ${{ vars.AGENT_TYPE }}
 
+      - id: get_cassettes_diff
+        name: Detect if cassettes changed
+        if: startsWith(github.event_name, 'pull_request')
+        run: |
+          cd tests/Auto-GPT-test-cassettes
+          cassette_diff=$(git diff --name-only --diff-filter=M -- '**/*.yaml')
+          
+          if [ -n "$cassette_diff" ]; then
+            echo "Cassettes changed: $cassette_diff"
+            echo "cassettes_changed=true" >> $GITHUB_OUTPUT
+          else
+            echo "Cassettes did not change."
+            echo "cassettes_changed=false" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Attempt to beat Challenges if cassettes changed
+        if: startsWith(github.event_name, 'pull_request') && steps.get_cassettes_diff.outputs.cassettes_changed == 'true'
+        run: |
+          pytest -n auto --beat-challenges --cov=autogpt --cov-branch --cov-report term-missing --cov-report xml \
+            tests/unit tests/integration tests/challenges
+          python tests/challenges/utils/build_current_score.py
+        env:
+          CI: true
+          PROXY: ${{ secrets.PROXY }}
+          AGENT_MODE: ${{ vars.AGENT_MODE }}
+          AGENT_TYPE: ${{ vars.AGENT_TYPE }}
+
       - name: Upload coverage reports to Codecov
         uses: codecov/codecov-action@v3
 
       - id: setup_git_auth
         name: Set up git token authentication
-        # Cassettes may be pushed even when tests fail
-        if: success() || failure()
+        # Cassettes may be pushed even when tests fail.
+        if: always() && (startsWith(github.event_name, 'pull_request') || success())
```

Comment:
this step is reached: 
- if previous steps are success: always
- if previous steps failed: only if PR

Reply:
This way it will continue running even when the workflow is canceled. That's what `always()` does and why I simplified the condition previously.
It's also hard to read because of the `always() &&` which is intuitively redundant.

Reply:
I didn't find a simpler way to write it.
success() || failure() is wrong because we don't want to push the cassettes on push event to master if the previous steps failed. That's very bad.

So I suggest keeping it like that.